brave/brave-core
# Fallback from Wireguard to IKEv2

Resolves https://github.com/brave/brave-browser/issues/30666

- Moved tunnel related utility functions to the tunnel level (added tunnel_utils.cc/h)
- Added tunnel usage flag:
  - Incremented when:
     - We failed to generate keypair using tunnel.dll
     - We setup vpn connection using tunnel.dll
  - Dropped on successful vpn disconnect.
- Added fallback dialog:
  - Detects bad state of wireguard vpn, failed attemts exceeded the limit.
  - Asks user to fallback to IKEv2, if yes, resets experiment flag and restarts the browser.
  - User can check the flag to skip this dialog if required.
  
![image](https://github.com/brave/brave-core/assets/2965009/31359de1-6597-44ab-9ee6-0c1de4ec77fb)




## Submitter Checklist:

  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`

## Reviewer Checklist:


## After-merge Checklist:

  changes has landed on

## Test Plan:

- case 1:
   - Install browser to program files
   - Enable wireguard experiment
   - Remove tunnel.dll or wireguard.dll and try to connect, after 3-4 attempts the browser should show the fallback dialog
 - case 2:
   - Install browser to program files
   - Enable wireguard experiment
   - Connect to VPN
   - Kill brave_vpn_wireguard_service.exe from Task Manager 
   - Repeat last 2 steps few times and you should see the fallback dialog.
